source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -12,70 +12,150 @@ namespace Extensions {
 namespace HttpFilters {
 namespace HeaderMutation {
 
-void Mutations::mutateRequestHeaders(Http::HeaderMap& headers,
-                                     const Formatter::HttpFormatterContext& ctx,
+void ParameterMutationAppend::mutateQueryParameter(
+    Http::Utility::QueryParamsMulti& params, const Formatter::HttpFormatterContext& context,
+    const StreamInfo::StreamInfo& stream_info) const {
+
+  const std::string value = formatter_->formatWithContext(context, stream_info);
+
+  switch (action_) {
+    PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
+  case ParameterAppendProto::APPEND_IF_EXISTS_OR_ADD:
+    params.add(key_, value);
+    return;
+  case ParameterAppendProto::ADD_IF_ABSENT: {
+    auto iter = params.data().find(key_);
+    if (iter == params.data().end()) {
+      params.add(key_, value);
+    }
+    break;
+  }
+  case ParameterAppendProto::OVERWRITE_IF_EXISTS: {
+    auto iter = params.data().find(key_);
+    if (iter != params.data().end()) {
+      params.overwrite(key_, value);
+    }
+    break;
+  }
+  case ParameterAppendProto::OVERWRITE_IF_EXISTS_OR_ADD:
+    params.overwrite(key_, value);
+    break;
+  }
+}
+
+Mutations::Mutations(const MutationsProto& config, absl::Status& creation_status) {
+  auto request_mutations_or_error = HeaderMutations::create(config.request_mutations());
+  SET_AND_RETURN_IF_NOT_OK(request_mutations_or_error.status(), creation_status);
+  request_mutations_ = std::move(request_mutations_or_error.value());
+
+  auto response_mutations_or_error = HeaderMutations::create(config.response_mutations());
+  SET_AND_RETURN_IF_NOT_OK(response_mutations_or_error.status(), creation_status);
+  response_mutations_ = std::move(response_mutations_or_error.value());
+
+  for (const auto& mutation : config.parameter_mutations()) {
+    if (mutation.has_append()) {
+      if (!mutation.append().has_record()) {
+        creation_status = absl::InvalidArgumentError("No record specified for parameter mutation.");
+        return;
+      }
+      if (!mutation.append().record().value().has_string_value()) {
```

Comment:
Yeah. If `value` is not set, the `record().value()` will return a value with `null` which will be rejected.

Reply:
> > Yeah. If `value` is not set, the `record().value()` will return a value with `null` which will be rejected.
> 
> I assume `record().value()` doesn't return null, because then this will seg-fault. Just to make sure this is addressed, please add a test.

Actually the `value()` here will always return a reference to `google::protobuf::Value`. If nothing is set, the content of the value would be null.

But a test always be better.